sdcm/scan_operation_thread.py

```diff
@@ -0,0 +1,352 @@
+                    self.time_elapsed = time.time() - start_time
+                    self.total_scan_time += self.time_elapsed
+                    self.log.info('[%s] last scan duration of %s rows is: %s', {type(self).__name__},
+                                  self.number_of_rows_read, self.time_elapsed)
+                    self.log.info('Average scan duration of %s scans is: %s', self.scans_counter,
+                                  (self.total_scan_time / self.scans_counter))
+                    operation_event.message = f"{type(self).__name__} operation ended successfully"
+                except Exception as exc:  # pylint: disable=broad-except
+                    msg = str(exc)
+                    msg = f"{msg} while running Nemesis: {db_node.running_nemesis}" if db_node.running_nemesis else msg
+                    operation_event.message = msg
+
+                    if db_node.running_nemesis or any(s in msg.lower() for s in ERROR_SUBSTRINGS):
+                        operation_event.severity = Severity.WARNING
+                    else:
+                        operation_event.severity = Severity.ERROR
+
+    def run_for_a_duration(self, scan_operation_event):
+        end_time = time.time() + self.duration
+        while time.time() < end_time and not self.termination_event.is_set():
+            self.db_node = random.choice(self.db_cluster.nodes)
+            self.read_pages = random.choice([100, 1000, 0])
+            self.scans_counter += 1
+            self.run_scan_operation(scan_operation_event=scan_operation_event)
+            self.log.info('Executed %s number: %s', scan_operation_event, self.scans_counter)
+            time.sleep(self.interval)
+
+    @abstractmethod
+    def run(self):
+        ...
+
+    def start(self):
+        self._thread.start()
+
+    def join(self, timeout=None):
+        return self._thread.join(timeout)
+
+
+class FullScanThread(ScanOperationThread):
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def randomly_form_cql_statement(self) -> Optional[str]:
+        cmd = self.randomly_bypass_cache(cmd=self.basic_query).format(self.ks_cf)
+        return self.randomly_add_timeout(cmd)
+
+    def run(self):
+        self.run_for_a_duration(scan_operation_event=FullScanEvent)
+
+
+class FullPartitionScanThread(ScanOperationThread):
+    """
+    Run a full scan of a partition, assuming it has a clustering key and multiple rows.
+    It runs a reversed query of a partition, then optionally runs a normal partition scan in order
+    to validate the reversed-query output data.
+
+    Should support the following query options:
+    1) ck < ?
+    2) ck > ?
+    3) ck > ? and ck < ?
+    4) order by ck desc
+    5) limit <int>
+    6) paging
+    """
+    reversed_query_filter_ck_by = {'lt': ' and {} < {}', 'gt': ' and {} > {}', 'lt_and_gt': ' and {} < {} and {} > {}',
+                                   'no_filter': ''}
+    reversed_query_order = ' order by ck desc'
```

Comment:
Doesn't the "desc" should be parameter?
Why then do you have get_table_clustering_order?

Reply:
I see below that there is a usage of "self.reversed_order", so probably this is an old piece of code.

Reply:
removed the old unused line.